File: bookstore-pos/backend/app/schemas/purchasing.py
from pydantic import BaseModel, Field


class PurchaseOrderItemCreate(BaseModel):
    product_id: int = Field(gt=0)
    qty: int = Field(gt=0)
    unit_cost: float = Field(ge=0)


class PurchaseOrderCreate(BaseModel):
    supplier_id: int = Field(gt=0)
    items: list[PurchaseOrderItemCreate] = Field(min_length=1)


class PurchaseOrderOut(BaseModel):
    id: int
    supplier_id: int
    status: str
    total: float
    model_config = {"from_attributes": True}


class PurchaseOrderItemOut(BaseModel):
    id: int
    purchase_order_id: int
    product_id: int
    qty: int
    unit_cost: float
    received_qty: int

    model_config = {"from_attributes": True}


class ReceiveItem(BaseModel):
    product_id: int = Field(gt=0)
    qty: int = Field(gt=0)


class ReceiveOrder(BaseModel):
    items: list[ReceiveItem] = Field(min_length=1)
    direct_costs_breakdown: dict[str, float] = Field(default_factory=dict)
    lot_prefix: str = "PO"


class SupplierPaymentCreate(BaseModel):
    supplier_id: int = Field(gt=0)
    amount: float = Field(gt=0)
    method: str = Field(min_length=1)
    reference: str = ""

File: bookstore-pos/frontend/src/modules/inventory/api/purchasing.ts
import { api } from "@/modules/shared/api";

export type PurchaseOrder = {
  id: number;
  supplier_id: number;
  status: string;
  total: number;
  created_at: string;
  items?: PurchaseOrderItem[];
};

export type PurchaseOrderItem = {
  id: number;
  purchase_order_id: number;
  product_id: number;
  qty: number;
  unit_cost: number;
  received_qty: number;
};

export const createPurchaseOrder = async (data: { supplier_id: number; items: { product_id: number; qty: number; unit_cost: number }[] }): Promise<PurchaseOrder> => {
  const res = await api.post("/purchasing/orders", data);
  return res.data;
};

export const listPurchaseOrders = async (status?: string): Promise<PurchaseOrder[]> => {
  const res = await api.get("/purchasing/orders", { params: status ? { status } : undefined });
  return res.data as PurchaseOrder[];
};

export const listPurchaseOrderItems = async (orderId: number): Promise<PurchaseOrderItem[]> => {
  const res = await api.get(`/purchasing/orders/${orderId}/items`);
  return res.data as PurchaseOrderItem[];
};

export const receivePurchaseOrder = async (orderId: number, items: { product_id: number; qty: number }[]) => {
  const res = await api.post(`/purchasing/orders/${orderId}/receive`, { items });
  return res.data;
};

export const receivePurchaseOrderWithCosts = async (
  orderId: number,
  payload: {
    items: { product_id: number; qty: number }[];
    direct_costs_breakdown?: Record<string, number>;
    lot_prefix?: string;
  }
) => {
  const res = await api.post(`/purchasing/orders/${orderId}/receive`, payload);
  return res.data;
};

export const supplierPayment = async (data: { supplier_id: number; amount: number; method: string; reference: string }) => {
  const res = await api.post("/purchasing/payments", data);
  return res.data;
};
